pkg/generate/cspc.go

```diff
@@ -0,0 +1,214 @@
+		Spec: cstorv1.CStorPoolClusterSpec{
+			Pools: *p,
+		},
+	}
+	// 6. Unmarshall it into a string
+	y, err := yaml.Marshal(cspc)
+	if err != nil {
+		fmt.Printf("err: %v\n", err)
+		return nil, "", err
+	}
+	yaml := string(y)
+	// 7. removing status and versionDetails field
+	yaml = yaml[:strings.Index(yaml, "status: {}")]
+	// 8. Split the string by the newlines/carriage returns and insert the BD's link
+	yaml = addBDDetailComments(yaml, bds)
+	return &cspc, yaml, nil
+}
+
+// addBDDetailComments adds more information about the blockdevice in a CSPC YAML string
+func addBDDetailComments(yaml string, bdList *v1alpha1.BlockDeviceList) string {
+	finalYaml := ""
+	for _, l := range strings.Split(yaml, "\n") {
+		if strings.Contains(l, "- blockDeviceName:") {
+			name := strings.Trim(strings.Split(l, ":")[1], " ")
+			finalYaml = finalYaml + getBDComment(name, bdList) + "\n"
+		}
+		finalYaml = finalYaml + l + "\n"
+	}
+	return finalYaml
+}
+
+// getBDComment returns information about a blockdevice, with fixed whitespace
+// to match the identation level
+func getBDComment(name string, bdList *v1alpha1.BlockDeviceList) string {
+	for _, bd := range bdList.Items {
+		if bd.Name == name {
+			return "      # " + bd.Spec.Path + "  " + util.ConvertToIBytes(strconv.FormatUint(bd.Spec.Capacity.Storage, 10))
+		}
+	}
+	return ""
+}
+
+// makePools creates a poolSpec based on the poolType, number of devices per
+// pool instance and a collection of blockdevices by nodes
+func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevice, nodes []string) (*[]cstorv1.PoolSpec, error) {
+	var spec []cstorv1.PoolSpec
+	if poolType == string(cstorv1.PoolStriped) {
+		// always a single RAID-group with nDevices patched together, cannot disk replace,
+		// no redundancy in a pool, redundancy possible across pool instances
+
+		// for each eligible set of BDs from each eligible node, take nDevices number of BDs
+		for _, node := range nodes {
+			bds := bd[node]
+			var raids []cstorv1.CStorPoolInstanceBlockDevice
+			for d := 0; d < nDevices; d++ {
+				raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[d].Name})
+			}
+			spec = append(spec, cstorv1.PoolSpec{
+				NodeSelector:   map[string]string{"kubernetes.io/hostname": node},
+				DataRaidGroups: []cstorv1.RaidGroup{{CStorPoolInstanceBlockDevices: raids}},
+				PoolConfig: cstorv1.PoolConfig{
+					DataRaidGroupType: string(cstorv1.PoolStriped),
+				},
+			})
+		}
+		return &spec, nil
+	} else if poolType == string(cstorv1.PoolMirrored) {
+		if nDevices%2 != 0 {
+			return nil, fmt.Errorf("mirrored pool requires multiples of two block device")
+		}
+		for node, bds := range bd {
+			var raids []cstorv1.CStorPoolInstanceBlockDevice
+			// add all BDs to a CSPCs CSPI spec
+			for d := 0; d < nDevices; d++ {
+				raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[d].Name})
+			}
```

Comment:
I have little concern over here... Let's say that the user is given to pick 4(nDevices) devices from a node to form a pool then I will suggest forming (nDevices/2) raidgroups instead of accumulating on same raidGroup

```go
   // Ignore this check if it already made in caller functions
   // Check if required bd's exist for given node
   if len(bds) < nDevices {
      return errors.New("required no.of(%d) blockdevices doesn't exist on node %s", nDevices, hostName)
   }
   bdIndex := 0
   var rgs []cstorv1.RaidGroup
  for rgIndex :=0; rgIndex<(nDevices/2); rgIndex++ {
          var raids []cstorv1.CStorPoolInstanceBlockDevice
          for d:=0; d<defaultMirrorCount; d++ {
                 raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[bdIndex].Name})
                 bdIndex++
          }
          rgs = append(rgs, raids)
  }
```
WDYT on above part?


Basically, with the current approach if the user configures 4 devices of size 10G then the size of the pool will be 10G(n redundancy will exist at pool level)
```sh
sparse-01d1417508c6ee62a915bfd0c341c78f        centos-worker-1   /var/openebs/sparse/2-ndm-sparse.img                10737418240      Unclaimed    Active     3h20m
sparse-07a7fbcd6a14cd09e21a539734ca4772        centos-worker-1   /var/openebs/sparse/4-ndm-sparse.img                10737418240      Unclaimed    Active     40s
sparse-5ac9c0a5b87adeecba382f3ec1c093d5        centos-worker-1   /var/openebs/sparse/1-ndm-sparse.img                10737418240      Unclaimed    Active     3h20m
sparse-9903f755f31c265bb190df228f2d3221        centos-worker-1   /var/openebs/sparse/5-ndm-sparse.img                10737418240      Unclaimed    Active     40s
sparse-f62a0627a72f86deba09e73f8f2dc809        centos-worker-1   /var/openebs/sparse/3-ndm-sparse.img                10737418240      Unclaimed    Active     3h20m

## Creating mirror pool with 4 blockdevices
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool create pool1 mirror /var/openebs/sparse/2-ndm-sparse.img  /var/openebs/sparse/4-ndm-sparse.img /var/openebs/sparse/1-ndm-sparse.img /var/openebs/sparse/5-ndm-sparse.img

### size of the pool is 10G
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zfs list
NAME                                         USED  AVAIL  REFER  MOUNTPOINT
pool1                                         53K  9.63G   512B  /pool1

### Pool vdev will looks like
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool status -v
  pool: pool1
 state: ONLINE
  scan: none requested
config:

	NAME                                      STATE     READ WRITE CKSUM
	pool1                                     ONLINE       0     0     0
	  mirror-0                                ONLINE       0     0     0
	    /var/openebs/sparse/2-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/4-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/1-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/5-ndm-sparse.img  ONLINE       0     0     0
```

With above suggestion it will be create pool of 20G with same configuration but redundancy of data will be only 1 copy than 3 copies
```sh
## Creating stripe pool on top of mirror pool
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool create pool1 mirror /var/openebs/sparse/2-ndm-sparse.img  /var/openebs/sparse/4-ndm-sparse.img mirror /var/openebs/sparse/1-ndm-sparse.img /var/openebs/sparse/5-ndm-sparse.img

## size of pool is 20G
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zfs list
NAME                                         USED  AVAIL  REFER  MOUNTPOINT
pool1                                         50K  19.3G   512B  /pool1

## pool vdev will looks like below
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool status
  pool: pool1
 state: ONLINE
  scan: none requested
config:

	NAME                                      STATE     READ WRITE CKSUM
	pool1                                     ONLINE       0     0     0
	  mirror-0                                ONLINE       0     0     0
	    /var/openebs/sparse/2-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/4-ndm-sparse.img  ONLINE       0     0     0
	  mirror-1                                ONLINE       0     0     0
	    /var/openebs/sparse/1-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/5-ndm-sparse.img  ONLINE       0     0     0

errors: No known data errors
```

Reply:
Okay, so you mean, I should cap the number of BDs in a RaidGroup to 2? If a user gives 6 `nDevices`, 3 RaidGroups should be formed in each pool instance spec?

Reply:
Got, it will do the same as suggested.

Reply:
[Update] I've done most of the work, I'm done stuck between this nodename & hostname quirk. Usually peeps would talk in terms of nodeName(because that's what they see in `kubectl get nodes`), so it'd make sense I show errors with nodeNames. Aah, this can get tricky, I need a map of `nodeName->hostName` after-all. 😞

Reply:
Wrote similar logic, since I know that the number of BDs in a RaidGroup will be fixed to `2`, i.e. default one, I'm just adding two BDs into the RaidGroup slice at a time.

Reply:
implemented similar logic to generate, RaidZ & RaidZ2 pools, I'll now have to do something to fix the capacity of the BDs which get selected in the pool assuming two cases
1. I have an infinite supply of BDs of different sizes
2. I have a finite supply of BDs and I should backtrack and re-select the most optimal ones which also minimize disk wastage